lib/ykpiv.c

```diff
@@ -704,6 +761,10 @@ ykpiv_rc ykpiv_verify(ykpiv_state *state, const char *pin, int *tries) {
   if((res = send_data(state, &apdu, data, &recv_len, &sw)) != YKPIV_OK) {
     return res;
   } else if(sw == SW_SUCCESS) {
+    if (state->pin == NULL) {
+      state->pin = malloc(strlen(pin) * sizeof(char));
```

Comment:
this allocates 1 byte to few to state->pin, strlen() excludes the \0 (additionally one might want to consider having state->pin be pre-allocated like char pin[9] and check with strlen() if it's set already.